### PR TITLE
d3d12: Be more robust with LUID-based adapter selection.

### LINKED ISSUE #1809: Wrong GPU being selected on hybrid graphics system
I am on a laptop that has both integrated graphics on the Intel CPU and a discrete NVIDIA GPU. When games are launched on Proton, they should be selecting the dGPU to run on, but when on D3D12 they consistently only use the integrated graphics.
`__NV_PRIME_RENDER_OFFLOAD` does not solve the issue, unfortunately. The discrete GPU still shows up at 0% usage in my system monitor and the game will run at less than 10 FPS.
This only occurs with D3D12 - when I run a game that uses Vulkan or D3D11, the NVIDIA GPU gets picked, as expected.

### Software information
I have tested this with Satisfactory (D3D12 vs. Vulkan rendering backends), Deep Rock Galactic (D3D12 vs. D3D11 rendering backends, and Forza Horizon 5 (only D3D12 and just won't run).

### System information
- GPU: Mesa Intel® Graphics and NVIDIA RTX 5000 Ada Generation Laptop GPU
- Driver: 4.18.1 and 545.29.06 respectively
- Wine version: Proton Experimental, but it also happens on 8.0.4
- VKD3D-Proton version: 2.11

### Log files
[steam-548430.log](https://github.com/HansKristian-Work/vkd3d-proton/files/13667371/steam-548430.log)

### PR DESCRIPTION
**Pull request by doitsujin**
Like the original PR (#1823), this should fix situations where multiple Vulkan adapters get assigned the same LUID, but with comparing device names this should also be robust against DXVK-side vendor ID overrides, which are still sometimes needed.

This is a situation that generally *shouldn't* happen and is probably some sort of winex11 issue in combination with systems that have multiple GPUs that can all output to the same display in one way or another.

Pinging @Etaash-mathamsetty , @TechnoPorg

Supersedes #1823.
Fixes #1809.

### CODE REVIEW

File: libs/d3d12core/main.c [hunk 1]
```diff
@@ -198,8 +199,33 @@ static VkPhysicalDevice d3d12_find_physical_device(struct vkd3d_instance *instan
...
+                {
+                    /* For simplicity, assume that adapter names are all ASCII characters */
+                    match = true;
+
+                    for (j = 0; j < ARRAY_SIZE(adapter_desc->Description); j++)
+                    {
+                        WCHAR a = (WCHAR)properties2.properties.deviceName[j];
+                        WCHAR b = adapter_desc->Description[j];
+
+                        if (!(match = (a == b)) || !a || !b)
```
Comment (HansKristian-Work) [inline libs/d3d12core/main.c:221]: This seems a bit iffy if string lengths don't match?

Comment (HansKristian-Work) [reply libs/d3d12core/main.c:221]: Might be better to add a wide string compare util to our string code?

Comment (doitsujin) [reply libs/d3d12core/main.c:221]: If string lengths don't match, then there's a point where one character is 0 but not the other, so a != b -> match = false. The last part wouldn't even be evaluated in that case.

Comment (HansKristian-Work) [reply libs/d3d12core/main.c:221]: Right ...

**Review by HansKristian-Work — APPROVED** [review]

**Comment by TechnoPorg** [comment]
Can confirm that this fixes the bug for me. Thanks a lot!

**Pull request merged**